smartsim/_core/utils/helpers.py

```diff
@@ -302,3 +306,43 @@ def decode_cmd(encoded_cmd: str) -> t.List[str]:
     cleaned_cmd = decoded_cmd.decode("ascii").split("|")
 
     return cleaned_cmd
+
+
+@t.final
+@dataclass(frozen=True)
+class _Pipeline(t.Generic[_T]):
+    """Utility class to turn
+
+    ..highlight:: python
+    ..code-block:: python
+
+        result = calls(function(nested(deeply(very(some(data))))))
+
+    into
+
+    ..highlight:: python
+    ..code-block:: python
+
+        result = (_Pipeline(data)
+                  .then(some)
+                  .then(very)
+                  .then(deeply)
+                  .then(nested)
+                  .then(function)
+                  .then(calls)
+                  .get_result())
+
+    without the need to introduce confusing temporary variable names
+    """
+
+    _val: _T
+
+    def then(self, fn: t.Callable[[_T], _U], /) -> "_Pipeline[_U]":
+        return _Pipeline(fn(self._val))
```

Comment:
I am not a huge fan of this. It seems like we start to mess with our logged call stacks and just make life harder.

I can also no longer put breakpoints to debug, either.... Are you sure this is desirable?

Reply:
That's a fair point! I made this to try and avoid the need for so many intermediate variables to the nature of
```py
request = ...
request_str = str(request)
request_schema = Schma.parse_obj(request_str)
```
and I didn't want to resort to
```py
calls(fn(nested(deeply(very(some(data))))))
```
but I think it ended up being less useful that I originally had hoped. I still like pipe-lining fns in but am willing to be out voted if others think its unreadable.

Reply:
Clearly, you are a lisp programmer at heart and should be feared and respected!